tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc

```diff
@@ -316,13 +316,23 @@ llvm::Value* EmitFullWarpShuffleDown(llvm::Value* value, llvm::Value* offset,
                                      llvm::IRBuilder<>* builder) {
   int bit_width = value->getType()->getPrimitiveSizeInBits();
   llvm::Value* all_warps_mask = builder->getInt32(-1);
+  llvm::Module* module = builder->GetInsertBlock()->getModule();
+  llvm::Triple target_triple = llvm::Triple(module->getTargetTriple());
 
   // Special case for efficiency
   if (value->getType()->isFloatTy() && bit_width == 32) {
-    return EmitCallToTargetIntrinsic(
-        TargetIntrinsicID::kShflDownF32,
-        {all_warps_mask, value, offset, builder->getInt32(kWarpSize - 1)}, {},
-        builder);
+    if (target_triple.getArch() == llvm::Triple::nvptx ||
+        target_triple.getArch() == llvm::Triple::nvptx64) {
+      return EmitShflDownIntrinsicFunction(
```

Comment:
I apologize. Part of my confusion is stemming from how much of target specific code to introduce into these emission files - also since I dont see any such instance in the existing code base.  I incorrectly assumed target_util will keep any such code and perhaps thats why the hesitation. I see that is not something you are concerned about at all.  Let me move the functionality to the same file which I agree allows for a simpler implementation.